[Pytorch Edge][tracing-based] build tracer in OSS

cccclai opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #66237
* #64152
* __->__ #64087
* #63421

1. Introduce
```
MobileModelRunner.h
MobileModelRunner.cpp
TensorUtils.h
TensorUtils.cpp
```
in external. They are pretty much the same as internal, except namespace and the dependency in folly. In next prs, TensorUtils and MobileModelRunner are unified between external and internal.
2. Introduce
```
tracer.cpp
```
for external. Majority is the same as internal one, with some cleanup on unnecessary dependency. It's unified between internal and external in next change.
3. Add an executable to build the tracer. It will be built for desktop only.



Differential Revision: [D30599136](https://our.internmc.facebook.com/intern/diff/D30599136/)

Differential Revision: [D30599136](https://our.internmc.facebook.com/intern/diff/D30599136)

Commit:
[Pytorch Edge][tracing-based] build tracer in OSS



Differential Revision: [D30599136](https://our.internmc.facebook.com/intern/diff/D30599136/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D30599136/)!

[ghstack-poisoned]

Commit:
Update on "[Pytorch Edge][tracing-based] build tracer in OSS"


Differential Revision: [D30599136](https://our.internmc.facebook.com/intern/diff/D30599136/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D30599136/)!

[ghstack-poisoned]

Commit:
Update on "[Pytorch Edge][tracing-based] build tracer in OSS"


Differential Revision: [D30599136](https://our.internmc.facebook.com/intern/diff/D30599136/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D30599136/)!

[ghstack-poisoned]

Commit:
Update on "[Pytorch Edge][tracing-based] build tracer in OSS"


Differential Revision: [D30599136](https://our.internmc.facebook.com/intern/diff/D30599136/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D30599136/)!

[ghstack-poisoned]

Commit:
Update on "[Pytorch Edge][tracing-based] build tracer in OSS"


Differential Revision: [D30599136](https://our.internmc.facebook.com/intern/diff/D30599136/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D30599136/)!

[ghstack-poisoned]

Commit:
Update on "[Pytorch Edge][tracing-based] build tracer in OSS"


Differential Revision: [D30599136](https://our.internmc.facebook.com/intern/diff/D30599136/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D30599136/)!

[ghstack-poisoned]

Commit:
Update on "[Pytorch Edge][tracing-based] build tracer in OSS"


Differential Revision: [D30599136](https://our.internmc.facebook.com/intern/diff/D30599136/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D30599136/)!

[ghstack-poisoned]

Commit:
Update on "[Pytorch Edge][tracing-based] build tracer in OSS"


Differential Revision: [D30599136](https://our.internmc.facebook.com/intern/diff/D30599136/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D30599136/)!

[ghstack-poisoned]

Commit:
Update on "[Pytorch Edge][tracing-based] build tracer in OSS"


Differential Revision: [D30599136](https://our.internmc.facebook.com/intern/diff/D30599136/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D30599136/)!

[ghstack-poisoned]

Commit:
Update on "[Pytorch Edge][tracing-based] build tracer in OSS"


Differential Revision: [D30599136](https://our.internmc.facebook.com/intern/diff/D30599136/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D30599136/)!

[ghstack-poisoned]

Commit:
Update on "[Pytorch Edge][tracing-based] build tracer in OSS"


Differential Revision: [D30599136](https://our.internmc.facebook.com/intern/diff/D30599136/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D30599136/)!

[ghstack-poisoned]

Commit:
Update on "[Pytorch Edge][tracing-based] build tracer in OSS"


Differential Revision: [D30599136](https://our.internmc.facebook.com/intern/diff/D30599136/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D30599136/)!

[ghstack-poisoned]

Commit:
Update on "[Pytorch Edge][tracing-based] build tracer in OSS"


1. Introduce
```
MobileModelRunner.h
MobileModelRunner.cpp
TensorUtils.h
TensorUtils.cpp
```
in external. They are pretty much the same as internal, except namespace and the dependency in folly. In next prs, TensorUtils and MobileModelRunner are unified between external and internal.
2. Introduce
```
tracer.cpp
```
for external. Majority is the same as internal one, with some cleanup on unnecessary dependency. It's unified between internal and external in next change.
3. Add an executable to build the tracer. It will be built for desktop only.



Differential Revision: [D30599136](https://our.internmc.facebook.com/intern/diff/D30599136/)

Differential Revision: [D30599136](https://our.internmc.facebook.com/intern/diff/D30599136)

[ghstack-poisoned]

Commit:
Update on "[Pytorch Edge][tracing-based] build tracer in OSS"


1. Introduce
```
MobileModelRunner.h
MobileModelRunner.cpp
TensorUtils.h
TensorUtils.cpp
```
in external. They are pretty much the same as internal, except namespace and the dependency in folly. In next prs, TensorUtils and MobileModelRunner are unified between external and internal.
2. Introduce
```
tracer.cpp
```
for external. Majority is the same as internal one, with some cleanup on unnecessary dependency. It's unified between internal and external in next change.
3. Add an executable to build the tracer. It will be built for desktop only.



Differential Revision: [D30599136](https://our.internmc.facebook.com/intern/diff/D30599136/)

Differential Revision: [D30599136](https://our.internmc.facebook.com/intern/diff/D30599136)

[ghstack-poisoned]

Commit:
Update on "[Pytorch Edge][tracing-based] build tracer in OSS"


1. Introduce
```
MobileModelRunner.h
MobileModelRunner.cpp
TensorUtils.h
TensorUtils.cpp
```
in external. They are pretty much the same as internal, except namespace and the dependency in folly. In next prs, TensorUtils and MobileModelRunner are unified between external and internal.
2. Introduce
```
tracer.cpp
```
for external. Majority is the same as internal one, with some cleanup on unnecessary dependency. It's unified between internal and external in next change.
3. Add an executable to build the tracer. It will be built for desktop only.



Differential Revision: [D30599136](https://our.internmc.facebook.com/intern/diff/D30599136/)

Differential Revision: [D30599136](https://our.internmc.facebook.com/intern/diff/D30599136)

[ghstack-poisoned]

Commit:
Update on "[Pytorch Edge][tracing-based] build tracer in OSS"


1. Introduce
```
MobileModelRunner.h
MobileModelRunner.cpp
TensorUtils.h
TensorUtils.cpp
```
in external. They are pretty much the same as internal, except namespace and the dependency in folly. In next prs, TensorUtils and MobileModelRunner are unified between external and internal.
2. Introduce
```
tracer.cpp
```
for external. Majority is the same as internal one, with some cleanup on unnecessary dependency. It's unified between internal and external in next change.
3. Add an executable to build the tracer. It will be built for desktop only.



Differential Revision: [D30599136](https://our.internmc.facebook.com/intern/diff/D30599136/)

Differential Revision: [D30599136](https://our.internmc.facebook.com/intern/diff/D30599136)

[ghstack-poisoned]

Commit:
Update on "[Pytorch Edge][tracing-based] build tracer in OSS"


1. Introduce
```
MobileModelRunner.h
MobileModelRunner.cpp
TensorUtils.h
TensorUtils.cpp
```
in external. They are pretty much the same as internal, except namespace and the dependency in folly. In next prs, TensorUtils and MobileModelRunner are unified between external and internal.
2. Introduce
```
tracer.cpp
```
for external. Majority is the same as internal one, with some cleanup on unnecessary dependency. It's unified between internal and external in next change.
3. Add an executable to build the tracer. It will be built for desktop only.



Differential Revision: [D30599136](https://our.internmc.facebook.com/intern/diff/D30599136/)

Differential Revision: [D30599136](https://our.internmc.facebook.com/intern/diff/D30599136)

[ghstack-poisoned]

Commit:
Update on "[Pytorch Edge][tracing-based] build tracer in OSS"


1. Introduce
```
MobileModelRunner.h
MobileModelRunner.cpp
TensorUtils.h
TensorUtils.cpp
```
in external. They are pretty much the same as internal, except namespace and the dependency in folly. In next prs, TensorUtils and MobileModelRunner are unified between external and internal.
2. Introduce
```
tracer.cpp
```
for external. Majority is the same as internal one, with some cleanup on unnecessary dependency. It's unified between internal and external in next change.
3. Add an executable to build the tracer. It will be built for desktop only.



Differential Revision: [D30599136](https://our.internmc.facebook.com/intern/diff/D30599136/)

Differential Revision: [D30599136](https://our.internmc.facebook.com/intern/diff/D30599136)

[ghstack-poisoned]

Commit:
Update on "[Pytorch Edge][tracing-based] build tracer in OSS"


1. Introduce
```
MobileModelRunner.h
MobileModelRunner.cpp
TensorUtils.h
TensorUtils.cpp
```
in external. They are pretty much the same as internal, except namespace and the dependency in folly. In next prs, TensorUtils and MobileModelRunner are unified between external and internal.
2. Introduce
```
tracer.cpp
```
for external. Majority is the same as internal one, with some cleanup on unnecessary dependency. It's unified between internal and external in next change.
3. Add an executable to build the tracer. It will be built for desktop only.



Differential Revision: [D30599136](https://our.internmc.facebook.com/intern/diff/D30599136/)

Differential Revision: [D30599136](https://our.internmc.facebook.com/intern/diff/D30599136)

[ghstack-poisoned]

Commit:
Update on "[Pytorch Edge][tracing-based] build tracer in OSS"


1. Introduce
```
MobileModelRunner.h
MobileModelRunner.cpp
TensorUtils.h
TensorUtils.cpp
```
in external. They are pretty much the same as internal, except namespace and the dependency in folly. In next prs, TensorUtils and MobileModelRunner are unified between external and internal.
2. Introduce
```
tracer.cpp
```
for external. Majority is the same as internal one, with some cleanup on unnecessary dependency. It's unified between internal and external in next change.
3. Add an executable to build the tracer. It will be built for desktop only.



Differential Revision: [D30599136](https://our.internmc.facebook.com/intern/diff/D30599136/)

Differential Revision: [D30599136](https://our.internmc.facebook.com/intern/diff/D30599136)

[ghstack-poisoned]

malfet commented:
Tentatively moving milestone to 1.10 as PR has not landed in master yet, although milestone was added more than a month ago

Commit:
Update on "[Pytorch Edge][tracing-based] build tracer in OSS"


1. Introduce
```
MobileModelRunner.h
MobileModelRunner.cpp
TensorUtils.h
TensorUtils.cpp
```
in external. They are pretty much the same as internal, except namespace and the dependency in folly. In next prs, TensorUtils and MobileModelRunner are unified between external and internal.
2. Introduce
```
tracer.cpp
```
for external. Majority is the same as internal one, with some cleanup on unnecessary dependency. It's unified between internal and external in next change.
3. Add an executable to build the tracer. It will be built for desktop only.



Differential Revision: [D30599136](https://our.internmc.facebook.com/intern/diff/D30599136/)

Differential Revision: [D30599136](https://our.internmc.facebook.com/intern/diff/D30599136)

[ghstack-poisoned]

malfet commented on torch/csrc/jit/mobile/model_tracer/CMakeLists.txt:
```diff
@@ -0,0 +1,18 @@
+set(
+  MODEL_TRACER_DIR
+  "${TORCH_ROOT}/torch/csrc/jit/mobile/model_tracer")
+set(MODEL_TRACER_DIR
+  ${TORCH_ROOT}/torch/csrc/jit/mobile/model_tracer/tracer.cpp
+  ${TORCH_ROOT}/torch/csrc/jit/mobile/model_tracer/TensorUtils.cpp
+  ${TORCH_ROOT}/torch/csrc/jit/mobile/model_tracer/MobileModelRunner.cpp
+  ${TORCH_ROOT}/torch/csrc/jit/mobile/model_tracer/OperatorCallTracer.cpp
+  ${TORCH_ROOT}/torch/csrc/jit/mobile/model_tracer/KernelDTypeTracer.cpp
+)
```
Please avoid filelists in CMakeFiles, use `build_variables.bzl` instead

cccclai replied:
Good point! Address in the latest commit.

Commit:
Update on "[Pytorch Edge][tracing-based] build tracer in OSS"


1. Introduce
```
MobileModelRunner.h
MobileModelRunner.cpp
TensorUtils.h
TensorUtils.cpp
```
in external. They are pretty much the same as internal, except namespace and the dependency in folly. In next prs, TensorUtils and MobileModelRunner are unified between external and internal.
2. Introduce
```
tracer.cpp
```
for external. Majority is the same as internal one, with some cleanup on unnecessary dependency. It's unified between internal and external in next change.
3. Add an executable to build the tracer. It will be built for desktop only.



Differential Revision: [D30599136](https://our.internmc.facebook.com/intern/diff/D30599136/)

Differential Revision: [D30599136](https://our.internmc.facebook.com/intern/diff/D30599136)

[ghstack-poisoned]

Commit:
Update on "[Pytorch Edge][tracing-based] build tracer in OSS"


1. Introduce
```
MobileModelRunner.h
MobileModelRunner.cpp
TensorUtils.h
TensorUtils.cpp
```
in external. They are pretty much the same as internal, except namespace and the dependency in folly. In next prs, TensorUtils and MobileModelRunner are unified between external and internal.
2. Introduce
```
tracer.cpp
```
for external. Majority is the same as internal one, with some cleanup on unnecessary dependency. It's unified between internal and external in next change.
3. Add an executable to build the tracer. It will be built for desktop only.



Differential Revision: [D30599136](https://our.internmc.facebook.com/intern/diff/D30599136/)

Differential Revision: [D30599136](https://our.internmc.facebook.com/intern/diff/D30599136)

[ghstack-poisoned]

Commit:
Update on "[Pytorch Edge][tracing-based] build tracer in OSS"


1. Introduce
```
MobileModelRunner.h
MobileModelRunner.cpp
TensorUtils.h
TensorUtils.cpp
```
in external. They are pretty much the same as internal, except namespace and the dependency in folly. In next prs, TensorUtils and MobileModelRunner are unified between external and internal.
2. Introduce
```
tracer.cpp
```
for external. Majority is the same as internal one, with some cleanup on unnecessary dependency. It's unified between internal and external in next change.
3. Add an executable to build the tracer. It will be built for desktop only.



Differential Revision: [D30599136](https://our.internmc.facebook.com/intern/diff/D30599136/)

Differential Revision: [D30599136](https://our.internmc.facebook.com/intern/diff/D30599136)

[ghstack-poisoned]

Commit:
Update on "[Pytorch Edge][tracing-based] build tracer in OSS"


1. Introduce
```
MobileModelRunner.h
MobileModelRunner.cpp
TensorUtils.h
TensorUtils.cpp
```
in external. They are pretty much the same as internal, except namespace and the dependency in folly. In next prs, TensorUtils and MobileModelRunner are unified between external and internal.
2. Introduce
```
tracer.cpp
```
for external. Majority is the same as internal one, with some cleanup on unnecessary dependency. It's unified between internal and external in next change.
3. Add an executable to build the tracer. It will be built for desktop only.



Differential Revision: [D30599136](https://our.internmc.facebook.com/intern/diff/D30599136/)

Differential Revision: [D30599136](https://our.internmc.facebook.com/intern/diff/D30599136)

[ghstack-poisoned]

Commit:
Update on "[Pytorch Edge][tracing-based] build tracer in OSS"


1. Introduce
```
MobileModelRunner.h
MobileModelRunner.cpp
TensorUtils.h
TensorUtils.cpp
```
in external. They are pretty much the same as internal, except namespace and the dependency in folly. In next prs, TensorUtils and MobileModelRunner are unified between external and internal.
2. Introduce
```
tracer.cpp
```
for external. Majority is the same as internal one, with some cleanup on unnecessary dependency. It's unified between internal and external in next change.
3. Add an executable to build the tracer. It will be built for desktop only.



Differential Revision: [D30599136](https://our.internmc.facebook.com/intern/diff/D30599136/)

Differential Revision: [D30599136](https://our.internmc.facebook.com/intern/diff/D30599136)

[ghstack-poisoned]

Commit:
Update on "[Pytorch Edge][tracing-based] build tracer in OSS"


1. Introduce
```
MobileModelRunner.h
MobileModelRunner.cpp
TensorUtils.h
TensorUtils.cpp
```
in external. They are pretty much the same as internal, except namespace and the dependency in folly. In next prs, TensorUtils and MobileModelRunner are unified between external and internal.
2. Introduce
```
tracer.cpp
```
for external. Majority is the same as internal one, with some cleanup on unnecessary dependency. It's unified between internal and external in next change.
3. Add an executable to build the tracer. It will be built for desktop only.



Differential Revision: [D30599136](https://our.internmc.facebook.com/intern/diff/D30599136/)

Differential Revision: [D30599136](https://our.internmc.facebook.com/intern/diff/D30599136)

[ghstack-poisoned]

Commit:
Update on "[Pytorch Edge][tracing-based] build tracer in OSS"


1. Introduce
```
MobileModelRunner.h
MobileModelRunner.cpp
TensorUtils.h
TensorUtils.cpp
```
in external. They are pretty much the same as internal, except namespace and the dependency in folly. In next prs, TensorUtils and MobileModelRunner are unified between external and internal.
2. Introduce
```
tracer.cpp
```
for external. Majority is the same as internal one, with some cleanup on unnecessary dependency. It's unified between internal and external in next change.
3. Add an executable to build the tracer. It will be built for desktop only.



Differential Revision: [D30599136](https://our.internmc.facebook.com/intern/diff/D30599136/)

Differential Revision: [D30599136](https://our.internmc.facebook.com/intern/diff/D30599136)

[ghstack-poisoned]

Commit:
Update on "[Pytorch Edge][tracing-based] build tracer in OSS"


1. Introduce
```
MobileModelRunner.h
MobileModelRunner.cpp
TensorUtils.h
TensorUtils.cpp
```
in external. They are pretty much the same as internal, except namespace and the dependency in folly. In next prs, TensorUtils and MobileModelRunner are unified between external and internal.
2. Introduce
```
tracer.cpp
```
for external. Majority is the same as internal one, with some cleanup on unnecessary dependency. It's unified between internal and external in next change.
3. Add an executable to build the tracer. It will be built for desktop only.



Differential Revision: [D30599136](https://our.internmc.facebook.com/intern/diff/D30599136/)

Differential Revision: [D30599136](https://our.internmc.facebook.com/intern/diff/D30599136)

[ghstack-poisoned]

Commit:
Update on "[Pytorch Edge][tracing-based] build tracer in OSS"


1. Introduce
```
MobileModelRunner.h
MobileModelRunner.cpp
TensorUtils.h
TensorUtils.cpp
```
in external. They are pretty much the same as internal, except namespace and the dependency in folly. In next prs, TensorUtils and MobileModelRunner are unified between external and internal.
2. Introduce
```
tracer.cpp
```
for external. Majority is the same as internal one, with some cleanup on unnecessary dependency. It's unified between internal and external in next change.
3. Add an executable to build the tracer. It will be built for desktop only.



Differential Revision: [D30599136](https://our.internmc.facebook.com/intern/diff/D30599136/)

Differential Revision: [D30599136](https://our.internmc.facebook.com/intern/diff/D30599136)

[ghstack-poisoned]
